firmware/src/rdm_responder.c

```diff
@@ -149,411 +120,239 @@ static void SendDUBResponseIfRequired(const uint8_t *param_data,
+  if (header->param_data_length) {
+    return RDMResponder_BuildNack(header, NR_FORMAT_ERROR);
   }
 
-  uint16_t param_id = ntohs(header->param_id);
-
-  if (header->command_class == DISCOVER_COMMAND) {
-    switch (param_id) {
-      case PID_DISC_UNIQUE_BRANCH:
-        SendDUBResponseIfRequired(param_data, param_data_length);
-        break;
-      case PID_DISC_MUTE:
-        Mute(header, param_data_length);
-        break;
-      case PID_DISC_UN_MUTE:
-        UnMute(header, param_data_length);
-        break;
-      default:
-        {}
-    }
-    return;
-  }
+  RDMResponder_BuildHeader(header, ACK, GET_COMMAND_RESPONSE,
+                           PID_IDENTIFY_DEVICE, sizeof(uint8_t));
+  g_rdm_buffer[sizeof(RDMHeader)] = g_responder.identify_on;
+  return RDMUtil_AppendChecksum(g_rdm_buffer);
+}
 
-  bool is_get = header->command_class == GET_COMMAND;
-  if (is_get && sub_device == SUBDEVICE_ALL) {
-    SendNack(header, NR_SUB_DEVICE_OUT_OF_RANGE);
+int RDMResponder_SetIdentifyDevice(const RDMHeader *header,
+                                   const uint8_t *param_data) {
+  if (header->param_data_length != sizeof(uint8_t)) {
+    return RDMResponder_BuildNack(header, NR_FORMAT_ERROR);
   }
-
-  switch (param_id) {
-    case PID_DEVICE_INFO:
-      if (is_get) {
-        GetDeviceInfo(header, param_data_length);
-      } else {
-        SendNack(header, NR_UNSUPPORTED_COMMAND_CLASS);
-      }
-      break;
-    case PID_SUPPORTED_PARAMETERS:
-      if (is_get) {
-        GetSupportedParameters(header, param_data_length);
-      } else {
-        SendNack(header, NR_UNSUPPORTED_COMMAND_CLASS);
-      }
-      break;
-    case PID_DEVICE_MODEL_DESCRIPTION:
-      if (is_get) {
-        GenericReturnString(header, param_data_length, DEVICE_MODEL_DESCRIPTION,
-                            sizeof(DEVICE_MODEL_DESCRIPTION));
-      } else {
-        SendNack(header, NR_UNSUPPORTED_COMMAND_CLASS);
-      }
-      break;
-    case PID_MANUFACTURER_LABEL:
-      if (is_get) {
-        GenericReturnString(header, param_data_length, MANUFACTURER_LABEL,
-                            sizeof(MANUFACTURER_LABEL));
-      } else {
-        SendNack(header, NR_UNSUPPORTED_COMMAND_CLASS);
-      }
-      break;
-    case PID_SOFTWARE_VERSION_LABEL:
-      if (is_get) {
-        GenericReturnString(header, param_data_length, SOFTWARE_LABEL,
-                            sizeof(SOFTWARE_LABEL));
-      } else {
-        SendNack(header, NR_UNSUPPORTED_COMMAND_CLASS);
-      }
+  switch (param_data[0]) {
+    case 0:
+      g_responder.identify_on = false;
       break;
-    case PID_IDENTIFY_DEVICE:
-      if (is_get) {
-        GetIdentifyDevice(header, param_data_length);
-      } else {
-        SetIdentifyDevice(header, param_data, param_data_length);
-      }
+    case 1:
+      g_responder.identify_on = true;
       break;
     default:
-      SendNack(header, NR_UNKNOWN_PID);
+      return RDMResponder_BuildNack(header, NR_DATA_OUT_OF_RANGE);
```

Comment:
The spec isn't that clear "Value for given Parameter out of allowable range or not supported." versus "The responder cannot interpret request as controller data was not formatted correctly.". I'd argue it depends if the field is a bool or uint8.